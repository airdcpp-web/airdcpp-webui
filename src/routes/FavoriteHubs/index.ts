export { default } from './components/FavoriteHubs';
