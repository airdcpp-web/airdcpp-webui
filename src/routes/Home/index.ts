import IconConstants from 'constants/IconConstants';

import * as UI from 'types/ui';

export const HOME_URL = '/home';

export const HomeRoutes: UI.RouteItem = {
  title: 'Home',
  path: HOME_URL,
  icon: IconConstants.HOME,
  lazy: async () => {
    const { default: Component } = await import(
      /* webpackChunkName: "home" */ './components'
    );
    return {
      Component,
    };
  },
};
