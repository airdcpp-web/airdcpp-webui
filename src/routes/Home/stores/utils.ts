import { saveWidgetSettings } from '@/routes/Home/widgets/WidgetUtils';
import { Layouts } from 'react-grid-layout';

import * as UI from '@/types/ui';
import {
  ApplicationWidgetInfo,
  ExtensionWidgetInfo,
  NotepadWidgetInfo,
  RSSWidgetInfo,
  TransferWidgetInfo,
} from '../widgets';
import { HomeLayoutColumns } from '../constants/HomeLayoutConstants';

export const createLayoutWidget = (
  layouts: Layouts,
  widgetInfo: UI.Widget,
  id: string,
  x?: number,
  y?: number,
) => {
  // Add the same widget for all layouts (we are lazy and use the same size for all layouts)
  return Object.keys(HomeLayoutColumns).reduce((reducedLayouts, key) => {
    reducedLayouts[key] = layouts[key] || [];
    reducedLayouts[key] = reducedLayouts[key].concat({
      i: id,
      x: x || (reducedLayouts[key].length * 2) % (HomeLayoutColumns[key] || 12),
      y: y || 0,
      ...widgetInfo.size,
    });

    return reducedLayouts;
  }, {} as Layouts);
};

const createDefaultLayoutWidget = <SettingsT>(
  layouts: Layouts,
  widgetInfo: UI.Widget,
  x: number,
  y: number,
  name?: string,
  settings?: SettingsT,
  suffix = '_default',
) => {
  const id = widgetInfo.typeId + suffix;
  const newLayout = createLayoutWidget(layouts, widgetInfo, id, x, y);

  saveWidgetSettings(id, {
    name,
    widget: settings,
  });

  return newLayout;
};

const createDefaultLayoutWidgets = (layouts: Layouts) => {
  layouts = createDefaultLayoutWidget(layouts, ApplicationWidgetInfo, 0, 0);
  layouts = createDefaultLayoutWidget(
    layouts,
    RSSWidgetInfo,
    ApplicationWidgetInfo.size.w,
    0,
    'News',
    {
      feed_url: 'https://airdcpp-web.github.io/feed.xml',
    },
    '_releases',
  );

  layouts = createDefaultLayoutWidget(
    layouts,
    ExtensionWidgetInfo,
    ApplicationWidgetInfo.size.w + RSSWidgetInfo.size.w,
    0,
  );
  layouts = createDefaultLayoutWidget(
    layouts,
    TransferWidgetInfo,
    ApplicationWidgetInfo.size.w + RSSWidgetInfo.size.w + ExtensionWidgetInfo.size.w,
    0,
  );
  layouts = createDefaultLayoutWidget(layouts, NotepadWidgetInfo, 0, 5);
  return layouts;
};

export const ensureDefaultLayoutWidgets = (layouts: Layouts) => {
  if (Object.keys(layouts).length !== 0) {
    return layouts;
  }

  // Initialize the default layout
  return createDefaultLayoutWidgets(layouts);
};
