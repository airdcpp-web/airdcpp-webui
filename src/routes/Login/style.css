.login-grid {
  margin: 0px !important;
  height: 100%;
}

.login-grid .column {
  max-width: 650px;
  width: 95% !important;
}
