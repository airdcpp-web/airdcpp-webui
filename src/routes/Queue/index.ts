export { default } from './components/Queue';
