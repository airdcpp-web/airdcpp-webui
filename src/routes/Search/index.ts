export { default } from './components/Search';
