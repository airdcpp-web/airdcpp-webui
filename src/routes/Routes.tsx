import * as React from 'react';

import { default as lazy } from '@/decorators/AsyncComponentDecorator';
import RouterMenuItemLink from '@/components/semantic/RouterMenuItemLink';

import LoginActions from '@/actions/reflux/LoginActions';
import IconConstants from '@/constants/IconConstants';

import * as API from '@/types/api';
import * as UI from '@/types/ui';

import { Trans } from 'react-i18next';
import { textToI18nKey } from '@/utils/TranslationUtils';
import { Route, matchPath, Location } from 'react-router';
import { APISocket } from '@/services/SocketService';

import { PrivateChatStoreSelector } from '@/stores/privateChatSessionSlice';
import { HubStoreSelector } from '@/stores/hubSessionSlice';
import { EventStoreSelector } from '@/stores/eventSlice';
import { FilelistStoreSelector } from '@/stores/filelistSlice';
import { ViewFileStoreSelector } from '@/stores/viewFileSlice';

export type RouteItemClickHandler = (
  path: string,
  event: React.SyntheticEvent<any>,
) => void;

export interface RouteItem {
  title: string;
  path: string;
  matchPath?: string;
  icon: string;
  unreadInfoStoreSelector?: UI.UnreadInfoStoreSelector;
  access?: API.AccessEnum;
  component?: React.ComponentType;
  className?: string;
  onClick?: RouteItemClickHandler;
}

export const HOME_URL = '/home';

export const mainRoutes: RouteItem[] = [
  {
    title: 'Home',
    path: HOME_URL,
    icon: IconConstants.HOME,
    component: lazy(() => import(/* webpackChunkName: "home" */ '@/routes/Home')),
  },
  {
    title: 'Queue',
    path: '/queue',
    icon: IconConstants.QUEUE_COLORED,
    access: API.AccessEnum.QUEUE_VIEW,
    component: lazy(() => import(/* webpackChunkName: "queue" */ '@/routes/Queue')),
  },
  {
    title: 'Search',
    path: '/search',
    icon: IconConstants.SEARCH,
    access: API.AccessEnum.SEARCH,
    component: lazy(() => import(/* webpackChunkName: "search" */ '@/routes/Search')),
  },
  {
    title: 'Transfers',
    path: '/transfers',
    icon: IconConstants.TRANSFERS_COLORED,
    access: API.AccessEnum.TRANSFERS,
    component: lazy(
      () => import(/* webpackChunkName: "transfers" */ '@/routes/Transfers'),
    ),
  },
];

export const configRoutes = [
  {
    title: 'Favorite hubs',
    path: '/favorite-hubs',
    icon: IconConstants.FAVORITE,
    access: API.AccessEnum.FAVORITE_HUBS_VIEW,
    component: lazy(
      () => import(/* webpackChunkName: "favorite-hubs" */ '@/routes/FavoriteHubs'),
    ),
  },
  {
    title: 'Share',
    path: '/share',
    icon: IconConstants.FOLDER,
    access: API.AccessEnum.SETTINGS_VIEW,
    component: lazy(() => import(/* webpackChunkName: "share" */ '@/routes/Share')),
  },
  {
    title: 'Settings',
    path: '/settings',
    // matchPath: '/settings/:mainSection?/:childSection?',
    icon: IconConstants.SETTINGS,
    access: API.AccessEnum.SETTINGS_VIEW,
    component: lazy(() => import(/* webpackChunkName: "settings" */ '@/routes/Settings')),
  },
];

export const secondaryRoutes: RouteItem[] = [
  {
    title: 'Hubs',
    path: '/hubs',
    matchPath: '/hubs/:session?/:id?',
    icon: IconConstants.HUBS_COLORED,
    unreadInfoStoreSelector: HubStoreSelector,
    access: API.AccessEnum.HUBS_VIEW,
    component: lazy(
      () => import(/* webpackChunkName: "hubs" */ '@/routes/Sidebar/routes/Hubs'),
    ),
  },
  {
    title: 'Messages',
    path: '/messages',
    matchPath: '/messages/:session?/:id?',
    icon: IconConstants.MESSAGES_COLORED,
    unreadInfoStoreSelector: PrivateChatStoreSelector,
    access: API.AccessEnum.PRIVATE_CHAT_VIEW,
    component: lazy(
      () => import(/* webpackChunkName: "messages" */ '@/routes/Sidebar/routes/Messages'),
    ),
  },
  {
    title: 'Filelists',
    path: '/filelists',
    matchPath: '/filelists/:session?/:id?',
    icon: IconConstants.FILELISTS_COLORED,
    unreadInfoStoreSelector: FilelistStoreSelector,
    access: API.AccessEnum.FILELISTS_VIEW,
    component: lazy(
      () =>
        import(/* webpackChunkName: "filelists" */ '@/routes/Sidebar/routes/Filelists'),
    ),
  },
  {
    title: 'Files',
    path: '/files',
    matchPath: '/files/:session?/:id?',
    icon: IconConstants.FILES_COLORED,
    unreadInfoStoreSelector: ViewFileStoreSelector,
    access: API.AccessEnum.VIEW_FILE_VIEW,
    component: lazy(
      () => import(/* webpackChunkName: "files" */ '@/routes/Sidebar/routes/Files'),
    ),
  },
  {
    title: 'Events',
    path: '/events',
    icon: IconConstants.EVENTS_COLORED,
    unreadInfoStoreSelector: EventStoreSelector,
    access: API.AccessEnum.EVENTS_VIEW,
    component: lazy(
      () => import(/* webpackChunkName: "system-log" */ '@/routes/Sidebar/routes/Events'),
    ),
  },
];

export const getLogoutItem = (socket: APISocket): RouteItem => ({
  icon: IconConstants.LOGOUT,
  path: 'logout',
  title: 'Logout',
  className: 'logout',
  onClick: (path, e) => {
    e.preventDefault();
    LoginActions.logout(socket);
  },
});

const menuItemClickHandler = (
  onClick: RouteItemClickHandler | undefined,
  route: RouteItem,
) => {
  if (!!onClick || !!route.onClick) {
    return (evt: any) =>
      onClick ? onClick(route.path, evt) : route.onClick!(route.path, evt);
  }

  return undefined;
};

export const parseMenuItem = (
  route: RouteItem,
  onClick: RouteItemClickHandler | undefined = undefined,
  showIcon: boolean | undefined = true,
) => {
  const { title, icon, unreadInfoStoreSelector, path, className } = route;
  return (
    <RouterMenuItemLink
      key={path}
      url={path}
      className={className}
      icon={showIcon ? icon + ' navigation' : null}
      onClick={menuItemClickHandler(onClick, route)}
      unreadInfoStoreSelector={unreadInfoStoreSelector}
    >
      <Trans i18nKey={textToI18nKey(title, UI.SubNamespaces.NAVIGATION)}>{title}</Trans>
    </RouterMenuItemLink>
  );
};

const filterItem = (item: RouteItem, { hasAccess }: UI.AuthenticatedSession) =>
  !item.access || hasAccess(item.access);

export const parseRoutes = (routes: RouteItem[]) => {
  return routes.map((route, i) => {
    const { component: Component, ...other } = route;
    return (
      <Route
        key={route.path}
        {...other}
        path={`${route.matchPath ? route.matchPath : route.path}/*`}
        element={!!Component && <Component />}
      />
    );
  });
};

export const parseMenuItems = (
  routes: RouteItem[],
  session: UI.AuthenticatedSession,
  onClick?: RouteItemClickHandler | undefined,
  showIcon?: boolean | undefined,
) => {
  return routes
    .filter((route) => filterItem(route, session))
    .map((route) => parseMenuItem(route, onClick, showIcon));
};

export const isRouteActive = (route: RouteItem, location: Location): boolean => {
  return !!matchPath(`${route.path}/*`, location.pathname);
};
