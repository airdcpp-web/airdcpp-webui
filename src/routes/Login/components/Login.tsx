import { useState, useRef } from 'react';
import * as React from 'react';

import LoginActions from 'actions/reflux/LoginActions';
import LoginStore from 'stores/LoginStore';

import Checkbox from 'components/semantic/Checkbox';
import { RouteComponentProps } from 'react-router';
import SocketConnectStatus from 'components/main/SocketConnectStatus';
import { useLoginState, LoginLocationState } from '../effects/UseLoginStateEffect';
import { ErrorBox, BottomMessage, SubmitButton } from './LoginLayoutComponents';

import '../style.css';
import { useTranslation } from 'react-i18next';
import { translate } from 'utils/TranslationUtils';

import * as UI from 'types/ui';
import Icon from 'components/semantic/Icon';
import IconConstants from 'constants/IconConstants';

interface LoginProps
  extends RouteComponentProps<UI.EmptyObject, any, LoginLocationState> {}

const Login: React.FC<LoginProps> = (props) => {
  const { t } = useTranslation();
  const [rememberMe, setRememberMe] = useState(false);
  const usernameRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);

  const [loading, setLoading] = useLoginState(props);
  if (!!LoginStore.refreshToken) {
    return (
      <SocketConnectStatus
        message={translate('Connecting to the server...', t, UI.Modules.LOGIN)}
      />
    );
  }

  const onSubmit = (evt: React.SyntheticEvent) => {
    const username = usernameRef.current!.value;
    const password = passwordRef.current!.value;
    evt.preventDefault();

    if (!username || !password) {
      return;
    }

    LoginActions.login(username, password, rememberMe);
    setLoading(true);
  };

  return (
    <div className="ui middle aligned center aligned grid login-grid">
      <div className="column">
        <form className="ui large form" autoComplete="on" onSubmit={onSubmit}>
          <div className="ui stacked segment">
            <div className="field">
              <div className="ui left icon input">
                <Icon icon={IconConstants.USER} />
                <input
                  ref={usernameRef}
                  type="text"
                  name="username"
                  placeholder={translate('Username', t, UI.Modules.LOGIN)}
                  required={true}
                  autoFocus={true}
                />
              </div>
            </div>
            <div className="field">
              <div className="ui left icon input">
                <Icon icon={IconConstants.LOCK} />
                <input
                  ref={passwordRef}
                  className="password"
                  name="password"
                  placeholder={translate('Password', t, UI.Modules.LOGIN)}
                  required={true}
                  type="password"
                />
              </div>
            </div>
            <SubmitButton
              onSubmit={onSubmit}
              loading={loading}
              allowLogin={LoginStore.allowLogin}
              t={t}
            />
            <div
              className="field"
              style={{
                marginTop: '10px',
                marginBottom: '3px',
              }}
            >
              <Checkbox
                caption={translate('Keep me logged in', t, UI.Modules.LOGIN)}
                onChange={setRememberMe}
                checked={rememberMe}
              />
            </div>
          </div>

          <BottomMessage />
        </form>

        <ErrorBox lastError={LoginStore.lastError} t={t} />
      </div>
    </div>
  );
};

export default Login;
