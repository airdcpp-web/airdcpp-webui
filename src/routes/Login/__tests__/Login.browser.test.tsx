import { waitFor } from '@testing-library/dom';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import { getMockServer, getSocket } from 'airdcpp-apisocket/tests';

import { renderBasicRoutes } from '@/tests/render/test-renderers';

import {
  setInputFieldValuesByPlaceholder,
  setupUserEvent,
} from '@/tests/helpers/test-form-helpers';
import Login from '../components/Login';
import {
  clickButton,
  expectResponseToMatchSnapshot,
  waitForUrl,
} from '@/tests/helpers/test-helpers';
import { getLogoutItem, parseMenuItem } from '@/routes/Routes';
import { useAppStore } from '@/context/AppStoreContext';
import { SocketContext } from '@/context/SocketContext';
import AuthenticationGuardDecorator from '@/components/main/decorators/AuthenticationGuardDecorator';
import {
  addMockSessionStoreInitDataHandlers,
  addMockSessionStoreSocketListeners,
} from '@/tests/mocks/mock-store';
import { useLayoutEffect } from 'react';
import { TestRouteNavigateButton } from '@/tests/helpers/test-route-helpers';
import { LOGIN_PROPS_KEY, REFRESH_TOKEN_KEY } from '@/stores/app/loginSlice';
import { saveLocalProperty, saveSessionProperty } from '@/utils/BrowserUtils';
import { getMockSession } from '@/tests/mocks/mock-session';

const AuthResponse = getMockSession([]);
const { refresh_token, ...ConnectResponse } = AuthResponse;

const ChildRouteUrl = '/child';
const ChildRouteCaption = 'Go to child';

// tslint:disable:no-empty
describe('Login', () => {
  let server: ReturnType<typeof getMockServer>;

  const addSuccessCredentialAuthHandlers = () => {
    const onLogin = vi.fn();
    const onLogout = vi.fn();

    server.addRequestHandler('POST', 'sessions/authorize', AuthResponse, onLogin);

    server.addRequestHandler('DELETE', 'sessions/self', undefined, onLogout);

    return { onLogin, onLogout };
  };

  const renderPage = async (initialRoute = '/login') => {
    const { socket } = getSocket({
      reconnectInterval: 0.1,
      logLevel: 'none',
      autoReconnect: false,
    });

    const SocketWrapper: React.FC<React.PropsWithChildren> = ({ children }) => (
      <SocketContext.Provider value={socket}>{children}</SocketContext.Provider>
    );

    const LoginPageTest = () => {
      return <Login />;
    };

    const LoggedInPageContent = AuthenticationGuardDecorator(() => {
      const appStore = useAppStore();
      return (
        <div>
          <div>Logged in</div>
          {parseMenuItem(getLogoutItem(socket, appStore))}
          <TestRouteNavigateButton route={ChildRouteUrl} caption={ChildRouteCaption} />
        </div>
      );
    });

    const LoggedInPage = () => {
      const appStore = useAppStore();
      useLayoutEffect(() => {
        addMockSessionStoreSocketListeners(server);
        addMockSessionStoreInitDataHandlers(server);
      }, [appStore.login.socketAuthenticated]);

      return <LoggedInPageContent />;
    };

    const routes = [
      {
        path: '/login/*',
        Component: LoginPageTest,
      },
      {
        index: true,
        Component: LoggedInPage,
      },
      {
        path: '/child/*',
        Component: LoggedInPage,
      },
    ];

    const renderData = renderBasicRoutes(
      routes,
      {
        routerProps: { initialEntries: [initialRoute] },
      },
      undefined,
      SocketWrapper,
    );

    const onSocketConnected = vi.fn();
    socket.onConnected = onSocketConnected;

    return { socket, onSocketConnected, ...renderData };
  };

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    localStorage.clear();
    sessionStorage.clear();

    server.stop();
  });

  const fillAndSubmitLogin = async ({
    getByRole,
    getByPlaceholderText,
    appStore,
  }: Awaited<ReturnType<typeof renderPage>>) => {
    // Check content
    await waitFor(() => expect(getByRole('form')).toBeTruthy(), {
      timeout: 10000,
    });

    // Edit
    const userEvent = setupUserEvent();
    await setInputFieldValuesByPlaceholder(
      { userEvent, getByPlaceholderText },
      {
        Username: 'testuser',
        Password: 'testpassword',
      },
    );

    // Submit
    const button = getByRole('button', { name: 'Login' });
    await userEvent.click(button);

    await waitFor(() =>
      expect(appStore.getState().login.socketAuthenticated).toBeTruthy(),
    );
  };

  const doInitialLogin = async () => {
    const authHandlerData = addSuccessCredentialAuthHandlers();

    const renderData = await renderPage();

    const { router, findAllByText } = renderData;
    await fillAndSubmitLogin(renderData);
    await waitForUrl('/', router);
    await waitFor(() => expect(findAllByText('Logged in')).toBeTruthy());

    return { ...renderData, ...authHandlerData };
  };

  describe('credentials', () => {
    test('should handle login', async () => {
      const userEvent = setupUserEvent();

      const { getByText, socket, onLogin, onLogout, onSocketConnected, appStore } =
        await doInitialLogin();

      await waitFor(() => expect(onSocketConnected).toHaveBeenCalledWith(AuthResponse));
      await waitFor(() => {
        expectResponseToMatchSnapshot(onLogin);
      });

      userEvent.click(getByText('Logout'));
      await waitFor(() => {
        expectResponseToMatchSnapshot(onLogout);
      });

      expect(socket.isConnected()).toBe(false);
      expect(appStore.getState().login.getSession()).toBeNull();
    }, 100000);

    test('should handle invalid credentials', async () => {
      const userEvent = setupUserEvent();

      const onLoginFailed = vi.fn();
      server.addErrorHandler(
        'POST',
        'sessions/authorize',
        'Invalid username or password',
        401,
        onLoginFailed,
      );

      const { getByRole, getByText, getByPlaceholderText } = await renderPage();

      // Check content
      await waitFor(() => expect(getByRole('form')).toBeTruthy());

      // Edit
      await setInputFieldValuesByPlaceholder(
        { userEvent, getByPlaceholderText },
        {
          Username: 'invalid',
          Password: 'invalid',
        },
      );

      // Submit
      const button = getByRole('button', { name: 'Login' });
      await userEvent.click(button);

      await waitFor(() => {
        expect(
          getByText('Authentication failed: Invalid username or password'),
        ).toBeTruthy();
      });
    }, 100000);
  });

  describe('reconnect', () => {
    test('should reconnect socket', async () => {
      const onReconnectSocket = vi.fn();

      const { getByRole, socket, appStore, router } = await doInitialLogin();

      clickButton(ChildRouteCaption, getByRole);
      await waitForUrl(ChildRouteUrl, router);

      // Disconnect socket
      socket.disconnect();

      await waitFor(() =>
        expect(appStore.getState().login.socketAuthenticated).toBeFalsy(),
      );

      // Let it reconnect
      server.addRequestHandler(
        'POST',
        'sessions/socket',
        ConnectResponse,
        onReconnectSocket,
      );

      await waitFor(() => expect(getByRole('progressbar')).toBeTruthy());

      await waitFor(() => {
        expect(appStore.getState().login.socketAuthenticated).toBeTruthy();
      });

      expectResponseToMatchSnapshot(onReconnectSocket);

      await waitForUrl(ChildRouteUrl, router);
    }, 100000);

    test('should handle reconnect errors', async () => {
      const onReconnectSocket = vi.fn();
      const onAuth = vi.fn();

      const renderData = await doInitialLogin();

      const { getByRole, socket, appStore, router } = renderData;

      clickButton(ChildRouteCaption, getByRole);
      await waitForUrl(ChildRouteUrl, router);

      saveLocalProperty(REFRESH_TOKEN_KEY, AuthResponse.refresh_token);

      socket.onConnect = () => {
        // Let it fail reconnect
        server.addErrorHandler(
          'POST',
          'sessions/socket',
          'Invalid session token',
          400,
          onReconnectSocket,
        );
      };

      // Disconnect socket
      socket.disconnect();

      await waitFor(() =>
        expect(appStore.getState().login.socketAuthenticated).toBeFalsy(),
      );

      await waitFor(() => expect(getByRole('progressbar')).toBeTruthy());
      await waitFor(() => expect(onReconnectSocket).toBeCalled());

      // We should be back on the login page
      await waitForUrl('/login', router);
      expect(appStore.getState().login.getSession()).toBeNull();

      // Refresh token should now be attempted
      server.addErrorHandler(
        'POST',
        'sessions/authorize',
        'Invalid refresh token',
        400,
        onAuth,
      );

      await waitFor(() => expect(onAuth).toBeCalled());

      // Refresh token should now be removed
      await waitFor(() => expect(appStore.getState().login.getRefreshToken()).toBeNull());

      // Re-login
      addSuccessCredentialAuthHandlers();
      await fillAndSubmitLogin(renderData);

      // The previously active page should be visible
      await waitForUrl(ChildRouteUrl, router);
    }, 100000);

    test('should connect with an existing session', async () => {
      saveSessionProperty(LOGIN_PROPS_KEY, AuthResponse);

      const onReconnectSocket = vi.fn();
      server.addRequestHandler(
        'POST',
        'sessions/socket',
        ConnectResponse,
        onReconnectSocket,
      );

      const { appStore, router } = await renderPage(ChildRouteUrl);

      await waitFor(() =>
        expect(appStore.getState().login.socketAuthenticated).toBeTruthy(),
      );

      expectResponseToMatchSnapshot(onReconnectSocket);

      await waitForUrl(ChildRouteUrl, router);
    }, 100000);
  });

  describe('refresh token', () => {
    test('should use refresh token with an expired session', async () => {
      const onReconnectSocket = vi.fn();
      saveSessionProperty(LOGIN_PROPS_KEY, AuthResponse);
      saveLocalProperty(REFRESH_TOKEN_KEY, AuthResponse.refresh_token);

      server.addErrorHandler(
        'POST',
        'sessions/socket',
        'Invalid session token',
        400,
        onReconnectSocket,
      );

      const onLogin = vi.fn();

      const { appStore, router, socket } = await renderPage(ChildRouteUrl);
      socket.onConnect = () => {
        // Handlers will be reset when the reconnect fails and socket gets disconnected
        server.addRequestHandler('POST', 'sessions/authorize', ConnectResponse, onLogin);
      };

      // Reconnect should fail
      await waitFor(() => expect(onReconnectSocket).toBeCalled());

      // We should be back on the login page
      await waitForUrl('/login', router);
      await waitFor(() =>
        expect(appStore.getState().login.socketAuthenticated).toBeTruthy(),
      );

      // Refresh token should be used for authorization and we are now back on the page that we originally wanted
      await waitForUrl(ChildRouteUrl, router);
      expectResponseToMatchSnapshot(onLogin);

      await waitForUrl(ChildRouteUrl, router);
    }, 100000);
  });
});
