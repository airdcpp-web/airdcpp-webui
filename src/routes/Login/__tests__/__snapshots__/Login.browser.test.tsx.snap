// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > credentials > should handle login 1`] = `
{
  "data": {
    "grant_type": "password",
    "password": "testpassword",
    "username": "testuser",
  },
  "method": "POST",
  "path": "sessions/authorize",
}
`;

exports[`Login > credentials > should handle login 2`] = `
{
  "method": "DELETE",
  "path": "sessions/self",
}
`;

exports[`Login > reconnect > should connect with an existing session 1`] = `
{
  "data": {
    "auth_token": "b823187f-4aab-4b71-9764-e63e88401a26",
  },
  "method": "POST",
  "path": "sessions/socket",
}
`;

exports[`Login > reconnect > should reconnect socket 1`] = `
{
  "data": {
    "auth_token": "b823187f-4aab-4b71-9764-e63e88401a26",
  },
  "method": "POST",
  "path": "sessions/socket",
}
`;

exports[`Login > refresh token > should use refresh token directly 1`] = `
{
  "data": {
    "grant_type": "refresh_token",
    "refresh_token": "dc66f47d-d060-4c50-aa3b-4a4741938a04",
  },
  "method": "POST",
  "path": "sessions/authorize",
}
`;

exports[`Login > refresh token > should use refresh token directly 2`] = `
{
  "data": {
    "grant_type": "refresh_token",
    "refresh_token": "dc66f47d-d060-4c50-aa3b-4a4741938a04",
  },
  "method": "POST",
  "path": "sessions/authorize",
}
`;

exports[`Login > refresh token > should use refresh token with an expired session 1`] = `
{
  "data": {
    "grant_type": "refresh_token",
    "refresh_token": "dc66f47d-d060-4c50-aa3b-4a4741938a04",
  },
  "method": "POST",
  "path": "sessions/authorize",
}
`;
